subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultProtectApiService.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks;
+
+import org.gradle.api.Action;
+import org.gradle.internal.exceptions.Contextual;
+
+public class DefaultProtectApiService implements ProtectApiService {
+    private int nestedActionExecutionCount = 0;
```

Comment:
given that this is a build scoped service, the implementation should be thread safe so that 2 threads can be running configuration actions for different projects concurrently. e.g. should use a `ThreadLocal` to track whether the current thread is configuring something or not, or perhaps make the service project scoped.

It would be good to add some test coverage for this case.

We should also clear the flag when crossing a project boundary via dependency resolution, so that if dependency resolution triggers configuration of the target project, the configuration logic for that project can call the hook methods if it so desires. Perhaps this can be a known issue for now and we fix it separately.

Reply:
I'm unsure here if we should disallow reaching over project boundary manually such as
```
project(':some-other-project') {
    tasks.register("foo") {
        rootProject.afterEvaluate { ... }
    }
}
```
Basically reaching in another project to add hooks? Or maybe we will just deprecate these via another way? By using a build service scope, I was able to prevent those project boundary reaching but it needs synchronization.

With regards to project boundary via dependency resolution, is this resolution handled in the same thread?